src/modelfree/density/fit_density.py

```diff
@@ -0,0 +1,317 @@
+    activation_glob = None  # directory of generated activatoins
+    output_root = None  # directory to write output
+    data_type = 'ff_policy'  # key into activations
+    max_timesteps = None  # if specified, maximum number of timesteps of activations to use
+    seed = 0
+    model_class = GaussianMixture  # density model to use
+    model_kwargs = {'n_components': 10}  # parameters for density model
+    train_opponent = 'zoo_1'  # opponent ID to use for fitting density model (extracted from path)
+    train_percentage = 0.5  # percentage of data to use for training (remainder is validation)
+    _ = locals()  # quieten flake8 unused variable warning
+    del _
+
+
+@fit_model_ex.named_config
+def debug_config():
+    max_timesteps = 100
+    model_kwargs = {'n_components': 2}
+
+    _ = locals()  # quieten flake8 unused variable warning
+    del _
+
+
+@fit_model_ex.named_config
+def gmm():
+    model_class = GaussianMixture
+    _ = locals()  # quieten flake8 unused variable warning
+    del _
+
+
+@fit_model_ex.named_config
+def pca_kde():
+    model_class = PCAPreDensity
+    model_kwargs = {'density_class': KernelDensity}
+    _ = locals()  # quieten flake8 unused variable warning
+    del _
+
+
+@fit_model_ex.named_config
+def pca_gmm():
+    model_class = PCAPreDensity
+    model_kwargs = {'density_class': GaussianMixture}
+    _ = locals()  # quieten flake8 unused variable warning
+    del _
+
+
+@fit_model_ex.named_config
+def kde():
+    model_class = KernelDensity
+    _ = locals()  # quieten flake8 unused variable warning
+    del _
+
+
+def _load_and_reshape_single_file(np_path, data_key, opponent_id):
+    """Loads data from np_path, extracting column col.
+
+    :param np_path: (str) a path to a pickled dict of NumPy arrays.
+    :param data_key: (str) the key to extract from the dict.
+    :param opponent_id: (str) an identifier, added to the metadata.
+    :return A tuple (concatenated_data, metadata_df) where concatenated_data is a flattened
+            array of activations at different timesteps, and metadata_df is a pandas DataFrame
+            containing the episode_id, timestep, frac_timestep and opponent_id.
+    """
+    traj_data = np.load(np_path, allow_pickle=True)
+    episode_list = traj_data[data_key].tolist()
+
+    episode_lengths = [len(episode) for episode in episode_list]
+    episode_id = []
+    timesteps = []
+    frac_timesteps = []
+    for i, episode_length in enumerate(episode_lengths):
+        episode_id += [i] * episode_length
+        episode_timesteps = list(range(episode_length))
+        timesteps += episode_timesteps
+        frac_timesteps += [el / episode_length for el in episode_timesteps]
+
+    concatenated_data = np.concatenate(episode_list)
+    opponent_id = [opponent_id] * len(concatenated_data)
+
+    # TODO: only opponent_id is used in this code. Are the others worth keeping around?
+    metadata_df = pd.DataFrame({'episode_id': episode_id,
```

Comment:
I'd enjoy keeping them around because we did consider e.g. trying to investigate the extent to which activations for the opponent or adversary differ at various places within an episode, even though we didn't have time to do that for this version of the paper

Reply:
thanks, I'll keep it for now